File: src/year2024/day1/day1.rb
def solve_part_1 lines
  left = []
  right = []

  lines.each do
    left_ele, right_ele = it.split
    left << left_ele.to_i
    right << right_ele.to_i
  end

  left.sort!
  right.sort!

  left
    .each_with_index
    .map { |ele, idx| (ele - right[idx]).abs }
    .sum
end

def solve_part_2 lines
  left = []
  right = []

  lines.each do
    left_ele, right_ele = it.split
    left << left_ele.to_i
    right << right_ele.to_i
  end

  left
    .map { |left_ele| left_ele * right.select { |right_ele| right_ele == left_ele }.size }
    .sum
end

require 'minitest/autorun'

class Day1Test < Minitest::Test
  def test_part_1_sample_input
    assert_equal(11, solve_part_1(File.readlines("test_input")))
  end

  def test_part_1_real_input
    assert_equal(1320851, solve_part_1(File.readlines("input")))
  end

  def test_part_2_sample_input
    assert_equal(31, solve_part_2(File.readlines("test_input")))
  end

  def test_part_2_real_input
    assert_equal(26859182, solve_part_2(File.readlines("input")))
  end
end

File: src/year2024/day1.rs
use itertools::Itertools;

pub fn solve_part_1(input: &str) -> usize {
    let mut left = vec![];
    let mut right = vec![];

    for line in input.lines() {
        let mut line_split = line.split_whitespace();
        let left_ele: usize = line_split.next().unwrap().parse().unwrap();
        let right_ele: usize = line_split.next().unwrap().parse().unwrap();

        left.push(left_ele);
        right.push(right_ele);
    }

    left.sort();
    right.sort();

    left.iter()
        .zip(right.iter())
        .map(|(a, b)| a.abs_diff(*b))
        .sum::<usize>()
}

pub fn solve_part_2(input: &str) -> usize {
    let mut left = vec![];
    let mut right = vec![];

    for line in input.lines() {
        let mut line_split = line.split_whitespace();
        let left_ele: usize = line_split.next().unwrap().parse().unwrap();
        let right_ele: usize = line_split.next().unwrap().parse().unwrap();

        left.push(left_ele);
        right.push(right_ele);
    }

    let tally = right.into_iter().counts();

    left.iter().map(|l| l * tally.get(l).or(Some(&0)).unwrap()).sum()
}

#[cfg(test)]
mod tests {
    #[test]
    fn part1_test_input() {
        assert_eq!(super::solve_part_1(&include_str!("day1/test_input")), 11);
    }

    #[test]
    fn part1_real_input() {
        assert_eq!(super::solve_part_1(&include_str!("day1/input")), 1320851);
    }

    #[test]
    fn part2_test_input() {
        assert_eq!(super::solve_part_2(&include_str!("day1/test_input")), 31);
    }

    #[test]
    fn part2_real_input() {
        assert_eq!(super::solve_part_2(&include_str!("day1/input")), 26859182);
    }
}
